Source/Particles/MultiParticleContainer.cpp

```diff
@@ -461,3 +464,247 @@ MultiParticleContainer::doContinuousInjection() const
     }
     return warpx_do_continuous_injection;
 }
+
+/* \brief Get ID of product species of each species.
+ * The users specifies the name of the product species, 
+ * this routine get its ID.
+ */
+void
+MultiParticleContainer::mapSpeciesProduct ()
+{
+    for (int i=0; i<nspecies; i++){
+        auto& pc = allcontainers[i];
+        // If species pc has ionization on, find species with name 
+        // pc->ionization_product_name and store its ID into 
+        // pc->ionization_product.
+        if (pc->do_field_ionization){
+            int i_product = getSpeciesID(pc->ionization_product_name);
+            AMREX_ALWAYS_ASSERT_WITH_MESSAGE(
+                i != i_product,
+                "ERROR: ionization product cannot be the same species");
+            pc->ionization_product = i_product;
+        }
+    }
+}
+
+/* \brief Given a species name, return its ID.
+ */
+int
+MultiParticleContainer::getSpeciesID (std::string product_str)
+{
+    int i_product;
+    bool found = 0;
+    // Loop over species
+    for (int i=0; i<nspecies; i++){
+        // If species name matches, store its ID
+        // into i_product
+        if (species_names[i] == product_str){
+            found = 1;
+            i_product = i;
+        }
+    }
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE(
+        found != 0,
+        "ERROR: could not find product species ID for ionization. Wrong name?");
+    return i_product;
+}
+
+namespace
+{
+    // For particle i in mfi, if is_ionized[i]=1, copy particle
+    // particle i from container pc_source into pc_product
+    static void createIonizedParticles (
```

Comment:
Why does this need to be `static void` and in a `namespace`?
Could this go into a separate file, for ionization?

Reply:
I removed the `static` keyword